bentoml/_internal/bento/build_config.py

```diff
@@ -138,58 +131,127 @@ def _envars_validator(
 
 @attr.frozen
 class DockerOptions:
+    """
+    .. note::
+
+        Refers to :ref:`concepts/bento:Docker Options` section for more information.
+
+    Args:
+        distro (:code:`str`, `optional`):
+            Distros to use in a 🍱 container.
+
+            .. note::
+
+                If :code:`None` is given, default to `debian`.
+
+        python_version (:code:`str`, `optional`):
+            Python version to use in a 🍱 container.
+
+            .. note::
+
+                If :code:`None` is given, default to current python version.
+
+        cuda_version (:code:`str| Literal["default"]`, `optional`):
+            CUDA version to use in a 🍱 container.
+
+            Currently supported CUDA versions are: 11.6.2
+
+            .. note::
+
+                :code:`cuda_version` also accepts :code:`"default"` to use the default version set by BentoML.
+                If you need supports for other CUDA versions, please open an issue at `BentoML Issue Tracker <https://github.com/bentoml/BentoML/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc>`_
+
+        env (:code:`dict[str, Any]`, `optional`):
+            A user-provided environment variable to be passed to a given bento.
+            We support the following formats:
+
+            - Passing as a YAML list of string:
+
+            .. code-block:: yaml
+
+                env:
+                - ENVAR=value1
+                - FOO=value2
+
+            - Passing as a YAML dictionary:
+
+            .. code-block:: yaml
+
+                env:
+                ENVAR: value1
+                FOO: value2
+
+            - Passing path to a :code:`.env` file:
+
+            .. code-block:: yaml
+
+                env: /path/to/.env
+
+            .. note ::
+
+                Make sure to follow the correct envar format: :code:`UPPERCASE=value`
+
+        system_packages (:code:`List[str]`, `optional`):
+            A user-provided system packages that can be install to a given bento based on :code:`distro` option.
+            BentoML will utilize the default system package manager for the given distro to install these packages.
+
+            .. note ::
+
+                System packages will be installed at  **Block** :code:`SETUP_BENTO_BASE_IMAGE`
+
+                **Block** is a component that construct the generated Dockerfile for a given bento.
+                Refers to :ref:`Dockerfile generation <reference/build_config:dockerfile generation>`.
+
+        setup_script (:code:`str`, `optional`):
+            A user-provided setup script to be executed during the construction of a 🍱 container.
+
+        base_image (:code:`str`, `optional`):
+            A user-provided base image to be used for a given 🍱 container.
+
+            .. note ::
+
+                If a base image is provided, :code:`distro`, :code:`python_version`, :code:`cuda_version` and :code:`system_packages` options will be ignored.
```

Comment:
The documentation here is great. I'm worried that if the argument mutual exclusion conditions get more complex, it will become harder to manager. Alternatively, we can introduce hierarchies of arguments, e.g. `from_distro`, `from_base_image`, to more naturally represent the exclusion conditions.  This is an areas I could learn more from you.

Reply:
I think by introducing hierarchies arguments, it would make the configuration even more complicated. The idea of a base image is that we want users to ensure that the provided images have the runtime dependency for bentoml, such as python, bash, cuda if they need to. We leave this responsibility to users solely to reduce the complexity on our end.